Sources/Core/AWSClientRuntime/Auth/CredentialsProviders/ECSCredentialsProvider.swift

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+import ClientRuntime
+import Foundation
+
+/// A credentials provider that sources credentials from ECS container metadata
+public struct ECSCredentialsProvider: CredentialsSourcedByCRT {
+    let crtCredentialsProvider: CRTCredentialsProvider
+    /// Creates a credential provider that sources credentials from ECS container metadata
+    /// ECS creds provider can be used to access creds via either relative uri to a fixed endpoint http://169.254.170.2,
+    /// or via a full uri specified by environment variables:
+    /// - AWS_CONTAINER_CREDENTIALS_RELATIVE_URI
+    /// - AWS_CONTAINER_CREDENTIALS_FULL_URI
+    /// - AWS_CONTAINER_AUTHORIZATION_TOKEN
+    ///
+    /// If both relative uri and absolute uri are set, relative uri has higher priority.
+    /// Token is used in auth header but only for absolute uri.
+    /// - Returns: `CredentialsProvider`
+    /// - Throws: CommonRuntimeError.crtError or InitializationError.missingURIs
+    public init(
+        environment: Environment = ProcessEnvironment()
+    ) throws {
+        let relativeURI = environment.environmentVariable(key: "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+        let absoluteURI = environment.environmentVariable(key: "AWS_CONTAINER_CREDENTIALS_FULL_URI")
+
+        guard relativeURI != nil || isValidAbsoluteURI(absoluteURI) else {
+            throw InitializationError.missingURIs(
+                "Please configure either the relative or absolute URI environment variable!"
+            )
+        }
+
+        let defaultHost = "169.254.170.2"
+        var host = defaultHost
+        var pathAndQuery = relativeURI ?? ""
+
+        if let absoluteString = absoluteURI, relativeURI == nil {
+            if let uriURL = URL(string: absoluteString) {
```

Comment:
I think combining these two using multiple let might make code simpler to read:

`if let absoluteString = absoluteURI, let uriURL = URL(string: absoluteString), relativeURI == nil { \\code }`

Reply:
I like the idea and in that case I will put `let uriURL = URL(string: absoluteString)` condition last since order here matters and I want to avoid any unnecessary evaluation of absoluteString if relativeURI is not nil.

Reply:
Yes, combining `if let` assignments with commas instead of nesting `if let`s is better.